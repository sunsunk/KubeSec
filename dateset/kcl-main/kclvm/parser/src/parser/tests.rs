use crate::parse_file_force_errors;
use regex::Regex;

#[test]
fn test_parse_file_not_found() {
    match parse_file_force_errors("The file path is invalid", None) {
        Ok(_) => {
            panic!("unreachable")
        }
        Err(err_msg) => {
            assert!(
                Regex::new(r"^Failed to load KCL file 'The file path is invalid'. Because.*")
                    .unwrap()
                    .is_match(&err_msg.to_string())
            );
        }
    }
}
