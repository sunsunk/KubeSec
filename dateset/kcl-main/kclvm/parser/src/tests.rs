use std::{
    env,
    panic::{catch_unwind, set_hook},
};

use compiler_base_span::{FilePathMapping, SourceMap};
use kclvm_config::modfile::{get_vendor_home, KCL_PKG_PATH};

use crate::*;

use core::any::Any;

mod ast;
mod error_recovery;
mod expr;
mod file;
mod types;

#[macro_export]
macro_rules! parse_expr_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_expr_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_module_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_module_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_type_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_type_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_type_node_snapshot {
    ($name:ident, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_type_node_string($src));
        }
    };
}

#[macro_export]
macro_rules! parse_file_ast_json_snapshot {
    ($name:ident, $filename:expr, $src:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_file_ast_json($filename, $src));
        }
    };
}

#[macro_export]
macro_rules! parse_file_snapshot {
    ($name:ident, $filename:expr) => {
        #[test]
        fn $name() {
            insta::assert_snapshot!($crate::tests::parsing_file_string($filename));
        }
    };
}

pub(crate) fn parsing_expr_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let expr = parser.parse_expr();
            format!("{expr:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub(crate) fn parsing_module_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sf = sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    match sf.src.as_ref() {
        Some(src_from_sf) => create_session_globals_then(|| {
            let stream = parse_token_streams(sess, src_from_sf.as_str(), new_byte_pos(0));
            let mut parser = Parser::new(sess, stream);
            let module = parser.parse_module();
            format!("{module:#?}\n")
        }),
        None => "".to_string(),
    }
}

pub(crate) fn parsing_type_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    create_session_globals_then(|| {
        let stream = parse_token_streams(sess, src, new_byte_pos(0));
        let mut parser = Parser::new(sess, stream);
        let typ = parser.parse_type_annotation();
        format!("{typ:#?}\n")
    })
}

pub(crate) fn parsing_type_node_string(src: &str) -> String {
    let sm = SourceMap::new(FilePathMapping::empty());
    sm.new_source_file(PathBuf::from("").into(), src.to_string());
    let sess = &ParseSession::with_source_map(Arc::new(sm));

    create_session_globals_then(|| {
        let stream = parse_token_streams(sess, src, new_byte_pos(0));
        let mut parser = Parser::new(sess, stream);
        let typ = parser.parse_type_annotation();
        typ.node.to_string()
    })
}

pub(crate) fn parsing_file_ast_json(filename: &str, src: &str) -> String {
    let m = crate::parse_file_with_global_session(
        Arc::new(ParseSession::default()),
        filename,
        Some(src.into()),
    )
    .unwrap();
    serde_json::ser::to_string_pretty(&m).unwrap()
}

pub(crate) fn parsing_file_string(filename: &str) -> String {
    let code = std::fs::read_to_string(filename).unwrap();
    let m = crate::parse_file(filename.trim_start_matches("testdata/"), Some(code))
        .expect(filename)
        .module;
    serde_json::ser::to_string_pretty(&m).unwrap()
}

pub fn check_result_panic_info(result: Result<(), Box<dyn Any + Send>>) {
    if let Err(e) = result {
        assert!(e.downcast::<String>().is_ok());
    };
}

const PARSE_EXPR_INVALID_TEST_CASES: &[&str] =
    &["fs1_i1re1~s", "fh==-h==-", "8_________i", "1MM", "0x00x"];

#[test]
pub fn test_parse_expr_invalid() {
    for case in PARSE_EXPR_INVALID_TEST_CASES {
        set_hook(Box::new(|_| {}));
        let result = catch_unwind(|| {
            parse_expr(case);
        });
        check_result_panic_info(result);
    }
}

const PARSE_FILE_INVALID_TEST_CASES: &[&str] = &[
    "a: int",                   // No initial value error
    "a -",                      // Invalid binary expression error
    "a?: int",                  // Invalid optional annotation error
    "if a not is not b: a = 1", // Logic operator error
    "if True:\n  a=1\n b=2",    // Indent error with recovery
    "a[1::::]",                 // List slice error
    "a[1 a]",                   // List index error
    "{a ++ 1}",                 // Config attribute operator error
    "func(a=1,b)",              // Call argument error
    "'${}'",                    // Empty string interpolation error
    "'${a: jso}'",              // Invalid string interpolation format spec error
];

#[test]
pub fn test_parse_file_invalid() {
    for case in PARSE_FILE_INVALID_TEST_CASES {
        let result = parse_file_force_errors("test.k", Some((&case).to_string()));
        assert!(result.is_err(), "case: {case}, result {result:?}");
    }
}

pub fn test_vendor_home() {
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCL_PKG_PATH, vendor);
    assert_eq!(get_vendor_home(), vendor.to_string());
}

fn set_vendor_home() -> String {
    // set env vendor
    let vendor = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap()
        .display()
        .to_string();
    env::set_var(KCL_PKG_PATH, vendor);
    debug_assert_eq!(get_vendor_home(), vendor.to_string());
    vendor.to_string()
}

#[test]
/// The testing will set environment variables,
/// so can not to execute test cases concurrently.
fn test_in_order() {
    test_import_vendor_by_external_arguments();
    println!("{:?} PASS", "test_import_vendor_by_external_arguments");
    test_import_vendor_without_vendor_home();
    println!("{:?} PASS", "test_import_vendor_without_vendor_home");
    test_import_vendor_without_kclmod();
    println!("{:?} PASS", "test_import_vendor_without_kclmod");
    test_import_vendor();
    println!("{:?} PASS", "test_import_vendor");
    test_import_vendor_with_same_internal_pkg();
    println!("{:?} PASS", "test_import_vendor_with_same_internal_pkg");
    test_import_vendor_without_kclmod_and_same_name();
    println!(
        "{:?} PASS",
        "test_import_vendor_without_kclmod_and_same_name"
    );
    test_vendor_home();
    println!("{:?} PASS", "test_vendor_home");
    test_pkg_not_found_suggestion();
    println!("{:?} PASS", "test_pkg_not_found_suggestion");
}

pub fn test_import_vendor() {
    let module_cache = KCLModuleCache::default();
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![
        ("assign.k", vec!["__main__", "assign", "assign.assign"]),
        (
            "config_expr.k",
            vec!["__main__", "config_expr", "config_expr.config_expr_02"],
        ),
        (
            "nested_vendor.k",
            vec![
                "__main__",
                "nested_vendor",
                "nested_vendor.nested_vendor",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
        (
            "subpkg.k",
            vec![
                "__main__",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub",
            ],
        ),
    ];

    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();

    let test_fn =
        |test_case_name: &&str, pkgs: &Vec<&str>, module_cache: Option<KCLModuleCache>| {
            let test_case_path = dir.join(test_case_name).display().to_string();
            let m = load_program(sess.clone(), &[&test_case_path], None, module_cache)
                .unwrap()
                .program;
            assert_eq!(m.pkgs.len(), pkgs.len());
            m.pkgs.into_iter().for_each(|(name, modules)| {
                println!("{:?} - {:?}", test_case_name, name);
                assert!(pkgs.contains(&name.as_str()));
                for pkg in pkgs.clone() {
                    if name == pkg {
                        if name == "__main__" {
                            assert_eq!(modules.len(), 1);
                            assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                        } else {
                            modules.into_iter().for_each(|module| {
                                assert!(module.filename.contains(&vendor));
                            });
                        }
                        break;
                    }
                }
            });
        };

    test_cases
        .iter()
        .for_each(|(test_case_name, pkgs)| test_fn(test_case_name, pkgs, None));

    test_cases.iter().for_each(|(test_case_name, pkgs)| {
        test_fn(test_case_name, pkgs, Some(module_cache.clone()))
    });
}

pub fn test_import_vendor_without_kclmod() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));

    let test_cases = vec![("import_vendor.k", vec!["__main__", "assign.assign"])];

    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .canonicalize()
        .unwrap();

    test_cases.into_iter().for_each(|(test_case_name, pkgs)| {
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None, None)
            .unwrap()
            .program;
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    });
}

pub fn test_import_vendor_without_vendor_home() {
    env::set_var(KCL_PKG_PATH, "");
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath assign not found in the program",
                "try 'kcl mod add assign' to download the package not found",
                "find more package on 'https://artifacthub.io'",
                "pkgpath assign.assign not found in the program",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }

    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath assign not found in the program",
                "try 'kcl mod add assign' to download the package not found",
                "find more package on 'https://artifacthub.io'",
                "pkgpath assign.assign not found in the program",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_with_same_internal_pkg() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata")
        .join("import_vendor")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("same_name.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `same_vendor` is found multiple times in the current package and vendor package"
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `same_vendor` is found multiple times in the current package and vendor package"
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_without_kclmod_and_same_name() {
    set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .join("same_name")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("assign.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `assign` is found multiple times in the current package and vendor package",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }

    match load_program(
        sess.clone(),
        &[&test_case_path],
        None,
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "the `assign` is found multiple times in the current package and vendor package",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}

fn test_import_vendor_by_external_arguments() {
    let vendor = set_vendor_home();
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let module_cache = KCLModuleCache::default();
    let external_dir = &PathBuf::from(".")
        .join("testdata")
        .join("test_vendor")
        .canonicalize()
        .unwrap();

    let test_cases = vec![
        (
            "import_by_external_assign.k",
            "assign",
            vec!["__main__", "assign"],
        ),
        (
            "import_by_external_config_expr.k",
            "config_expr",
            vec!["__main__", "config_expr"],
        ),
        (
            "import_by_external_nested_vendor.k",
            "nested_vendor",
            vec![
                "__main__",
                "nested_vendor",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
        (
            "import_by_external_vendor_subpkg.k",
            "vendor_subpkg",
            vec![
                "__main__",
                "vendor_subpkg",
                "vendor_subpkg.sub.sub1",
                "vendor_subpkg.sub.sub2",
                "vendor_subpkg.sub.sub",
                "vendor_subpkg.sub",
            ],
        ),
    ];

    let dir = &PathBuf::from(".")
        .join("testdata_without_kclmod")
        .canonicalize()
        .unwrap();

    let test_fn = |test_case_name: &&str,
                   dep_name: &&str,
                   pkgs: &Vec<&str>,
                   module_cache: Option<KCLModuleCache>| {
        let mut opts = LoadProgramOptions::default();
        opts.package_maps.insert(
            dep_name.to_string(),
            external_dir.join(dep_name).display().to_string(),
        );
        let test_case_path = dir.join(test_case_name).display().to_string();
        let m = load_program(sess.clone(), &[&test_case_path], None, module_cache)
            .unwrap()
            .program;
        assert_eq!(m.pkgs.len(), pkgs.len());
        m.pkgs.into_iter().for_each(|(name, modules)| {
            assert!(pkgs.contains(&name.as_str()));
            for pkg in pkgs.clone() {
                if name == pkg {
                    if name == "__main__" {
                        assert_eq!(modules.len(), 1);
                        assert_eq!(modules.get(0).unwrap().filename, test_case_path);
                    } else {
                        modules.into_iter().for_each(|module| {
                            assert!(module.filename.contains(&vendor));
                        });
                    }
                    break;
                }
            }
        });
    };

    test_cases
        .iter()
        .for_each(|(test_case_name, dep_name, pkgs)| test_fn(test_case_name, dep_name, pkgs, None));

    test_cases
        .iter()
        .for_each(|(test_case_name, dep_name, pkgs)| {
            test_fn(test_case_name, dep_name, pkgs, Some(module_cache.clone()))
        });
}

#[test]
fn test_get_compile_entries_from_paths() {
    let testpath = PathBuf::from("./src/testdata/multimods")
        .canonicalize()
        .unwrap();

    // [`kcl1_path`] is a normal path of the package [`kcl1`] root directory.
    // It looks like `/xxx/xxx/xxx`.
    let kcl1_path = testpath.join("kcl1");

    // [`kcl2_path`] is a mod relative path of the packege [`kcl2`] root directory.
    // It looks like `${kcl2:KCL_MOD}/xxx/xxx`
    let kcl2_path = PathBuf::from("${kcl2:KCL_MOD}/main.k");

    // [`kcl3_path`] is a mod relative path of the [`__main__`] packege.
    let kcl3_path = PathBuf::from("${KCL_MOD}/main.k");

    // [`package_maps`] is a map to show the real path of the mod relative path [`kcl2`].
    let mut opts = LoadProgramOptions::default();
    opts.package_maps.insert(
        "kcl2".to_string(),
        testpath.join("kcl2").to_str().unwrap().to_string(),
    );

    // [`get_compile_entries_from_paths`] will return the map of package name to package root real path.
    let entries = get_compile_entries_from_paths(
        &[
            kcl1_path.to_str().unwrap().to_string(),
            kcl2_path.display().to_string(),
            kcl3_path.display().to_string(),
        ],
        &opts,
    )
    .unwrap();

    assert_eq!(entries.len(), 3);

    assert_eq!(entries.get_nth_entry(0).unwrap().name(), "__main__");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(0).unwrap().path())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        kcl1_path.canonicalize().unwrap().to_str().unwrap()
    );

    assert_eq!(entries.get_nth_entry(1).unwrap().name(), "kcl2");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(1).unwrap().path())
            .canonicalize()
            .unwrap()
            .display()
            .to_string(),
        testpath
            .join("kcl2")
            .canonicalize()
            .unwrap()
            .to_str()
            .unwrap()
    );

    assert_eq!(entries.get_nth_entry(2).unwrap().name(), "__main__");
    assert_eq!(
        PathBuf::from(entries.get_nth_entry(2).unwrap().path())
            .canonicalize()
            .unwrap()
            .to_str()
            .unwrap(),
        kcl1_path.canonicalize().unwrap().to_str().unwrap()
    );
}

#[test]
fn test_dir_with_k_code_list() {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let testpath = PathBuf::from("./src/testdata/test_k_code_list")
        .canonicalize()
        .unwrap();

    let mut opts = LoadProgramOptions::default();
    opts.k_code_list = vec!["test_code = 1".to_string()];

    match load_program(
        sess.clone(),
        &[&testpath.display().to_string()],
        Some(opts.clone()),
        None,
    ) {
        Ok(_) => panic!("unreachable code"),
        Err(err) => assert!(err.to_string().contains("Invalid code list")),
    }

    match load_program(
        sess.clone(),
        &[&testpath.display().to_string()],
        Some(opts),
        Some(KCLModuleCache::default()),
    ) {
        Ok(_) => panic!("unreachable code"),
        Err(err) => assert!(err.to_string().contains("Invalid code list")),
    }
}

pub fn test_pkg_not_found_suggestion() {
    let sm = SourceMap::new(FilePathMapping::empty());
    let sess = Arc::new(ParseSession::with_source_map(Arc::new(sm)));
    let dir = &PathBuf::from("./src/testdata/pkg_not_found")
        .canonicalize()
        .unwrap();
    let test_case_path = dir.join("suggestions.k").display().to_string();
    match load_program(sess.clone(), &[&test_case_path], None, None) {
        Ok(_) => {
            let errors = sess.classification().0;
            let msgs = [
                "pkgpath k9s not found in the program",
                "try 'kcl mod add k9s' to download the package not found",
                "find more package on 'https://artifacthub.io'",
            ];
            assert_eq!(errors.len(), msgs.len());
            for (diag, m) in errors.iter().zip(msgs.iter()) {
                assert_eq!(diag.messages[0].message, m.to_string());
            }
        }
        Err(_) => {
            panic!("Unreachable code.")
        }
    }
}
