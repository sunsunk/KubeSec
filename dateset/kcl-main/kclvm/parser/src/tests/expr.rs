use crate::tests::parse_expr_snapshot;

parse_expr_snapshot!(smoke_test_parsing_expr_0, "1\n");
parse_expr_snapshot!(smoke_test_parsing_expr_1, "\"1\"\n");
parse_expr_snapshot!(named_literal_expr_0, "Undefined");
parse_expr_snapshot!(named_literal_expr_1, "None");
parse_expr_snapshot!(named_literal_expr_2, "True");
parse_expr_snapshot!(named_literal_expr_3, "False");
parse_expr_snapshot!(nonstring_literal_expr, r####"1234"####);
parse_expr_snapshot!(string_literal_expr_0, r####"'1234'"####);
parse_expr_snapshot!(string_literal_expr_1, r####""1234""####);
parse_expr_snapshot!(string_literal_expr_2, r####""1234\n""####);
parse_expr_snapshot!(number_bin_suffix_expr, r####"1234Ki"####);
parse_expr_snapshot!(unary_expr, r####"+1"####);
parse_expr_snapshot!(binary_expr_0, r####"1+2+3"####);
parse_expr_snapshot!(binary_expr_1, r####"1+2*3-4"####);
parse_expr_snapshot!(binary_expr_2, r####"1+2*3/4"####);
parse_expr_snapshot!(binary_expr_3, r####"a or b"####);
parse_expr_snapshot!(binary_expr_4, r####"x == a or b"####);
parse_expr_snapshot!(binary_expr_5, r####"22 > 11 and 111 < 222"####);
parse_expr_snapshot!(binary_expr_6, r####"int(e.value) > 1 and i == 0"####);
parse_expr_snapshot!(binary_expr_7, r####"key in ['key']"####);
parse_expr_snapshot!(binary_expr_8, r####"key not in ['key']"####);
parse_expr_snapshot!(binary_expr_9, r####"1 is 1 and 11 is not 22"####);
parse_expr_snapshot!(binary_expr_10, r####"1 + a and b"####);
parse_expr_snapshot!(binary_expr_with_paren, r####"1*(2+3)-4"####);
parse_expr_snapshot!(logic_expr_0, r####"0 < a < 100"####);
parse_expr_snapshot!(logic_expr_1, r####"0 < a < 100 + a"####);
parse_expr_snapshot!(logic_expr_2, r####"100 > a > 0"####);
parse_expr_snapshot!(logic_expr_3, r####"100 + a > a > 0"####);
parse_expr_snapshot!(logic_expr_4, r####"a is b"####);
parse_expr_snapshot!(logic_expr_5, r####"a is not True"####);
parse_expr_snapshot!(logic_expr_6, r####"not False or a > 0 and b is True"####);
parse_expr_snapshot!(if_expr, r####"1 if true else 2"####);
parse_expr_snapshot!(primary_expr_0, r####"a.b.c"####);
parse_expr_snapshot!(primary_expr_1, r####"'{}'.format(1)"####);
parse_expr_snapshot!(primary_expr_2, r####"str(1).isdigit()"####);
parse_expr_snapshot!(list_expr_0, r####"[1, 2, 3]"####);
parse_expr_snapshot!(list_expr_1, r####"[1, if True: 2, 3]"####);
parse_expr_snapshot!(list_comp_expr_0, r####"[x ** 2 for x in [1, 2, 3]]"####);
parse_expr_snapshot!(list_comp_expr_1, r####"[i for i in [1, 2, 3] if i > 2]"####);
parse_expr_snapshot!(dict_expr, r####"{k0=v0, k1=v1}"####);
parse_expr_snapshot!(
    dict_comp_expr,
    r####"{k: v + 1 for k, v in {k1 = 1, k2 = 2}}"####
);
parse_expr_snapshot!(subscript_expr_0, r####"a[0]"####);
parse_expr_snapshot!(subscript_expr_1, r####"b["k"]"####);
parse_expr_snapshot!(subscript_expr_2, r####"c?[1]"####);
parse_expr_snapshot!(subscript_expr_3, r####"a[1:]"####);
parse_expr_snapshot!(subscript_expr_4, r####"a[:-1]"####);
parse_expr_snapshot!(subscript_expr_5, r####"a[1:len]"####);
parse_expr_snapshot!(subscript_expr_6, r####"a[0:-1]"####);
parse_expr_snapshot!(subscript_expr_7, r####"a[::]"####);
parse_expr_snapshot!(subscript_expr_8, r####"a[1::]"####);
parse_expr_snapshot!(subscript_expr_9, r####"a[:0:]"####);
parse_expr_snapshot!(subscript_expr_10, r####"a[::-1]"####);
parse_expr_snapshot!(subscript_expr_11, r####"a[1::2]"####);
parse_expr_snapshot!(subscript_expr_12, r####"a[:2:1]"####);
parse_expr_snapshot!(subscript_expr_13, r####"a[1:2:]"####);
parse_expr_snapshot!(subscript_expr_14, r####"a[1:3:1]"####);
parse_expr_snapshot!(call_expr_0, r####"func0()"####);
parse_expr_snapshot!(call_expr_1, r####"func1(1)"####);
parse_expr_snapshot!(call_expr_2, r####"func2(x=2)"####);
parse_expr_snapshot!(call_expr_3, r####"func3(1,x=2)"####);
parse_expr_snapshot!(quant_expr_0, r####"all x in collection {x > 0}"####);
parse_expr_snapshot!(quant_expr_1, r####"any y in collection {y < 0}"####);
parse_expr_snapshot!(quant_expr_2, r####"map x in collection {x + 1}"####);
parse_expr_snapshot!(quant_expr_3, r####"filter x in collection {x > 1}"####);
parse_expr_snapshot!(quant_expr_4, r####"filter x in collection {x > 1}"####);
parse_expr_snapshot!(
    quant_expr_5,
    r####"map i, e in [{k1 = "v1", k2 = "v2"}] { e }"####
);
parse_expr_snapshot!(
    quant_expr_6,
    r####"map i, e in [{k1 = "v1", k2 = "v2"}] { e if i > 0 }"####
);
parse_expr_snapshot!(lambda_expr_0, r####"lambda {}"####);
parse_expr_snapshot!(lambda_expr_1, r####"lambda x {}"####);
parse_expr_snapshot!(lambda_expr_2, r####"lambda x: int -> int {x}"####);
parse_expr_snapshot!(
    lambda_expr_3,
    r####"lambda {
    if True:
        _a = 1
    else:
        _a = 2
    _a
}"####
);
parse_expr_snapshot!(
    config_expr_0,
    r####"{
    "name" = {
        "name": "alice"
    },
    "gender" = "female"
}"####
);
parse_expr_snapshot!(
    config_expr_1,
    r####"{
    "name" = {
        "name": "alice"
    }
    "gender" = "female",
}"####
);
parse_expr_snapshot!(
    config_expr_2,
    r####"{
    "name" = {
        "name": "alice",
    }
    "gender" = "female"
}"####
);
parse_expr_snapshot!(
    config_if_expr_0,
    r####"{
    if True:
        a = 1
}"####
);
parse_expr_snapshot!(
    config_if_expr_1,
    r####"{
    if True:
        a = 1
    else:
        a = 2
}"####
);
parse_expr_snapshot!(
    config_if_expr_2,
    r####"{
    if True:
        a = 1
    elif x > 1:
        a = 2
    else:
        a = 3
}"####
);
parse_expr_snapshot!(
    config_if_expr_3,
    r####"{
    if True:
        if False:
            a = 1
}"####
);
parse_expr_snapshot!(schema_expr_0, r####"Schema {}"####);
parse_expr_snapshot!(schema_expr_1, r####"Schema {k=v}"####);
parse_expr_snapshot!(schema_expr_2, r####"Schema () {k=v}"####);
parse_expr_snapshot!(schema_expr_3, r####"Schema (1, 2) {k=v}"####);
parse_expr_snapshot!(
    schema_expr_4,
    r####"Schema (1, 2) {
    k=v
}"####
);
parse_expr_snapshot!(
    line_continue,
    r####"1 + \
2
"####
);
parse_expr_snapshot!(parse_joined_string_0, r####"'${123+200}'"####);
parse_expr_snapshot!(parse_joined_string_1, r####"'abc${a+1}cde'"####);
parse_expr_snapshot!(expr_with_paren_0, r####"(2+3)"####);
parse_expr_snapshot!(expr_with_paren_1, r####"((2+3)"####);
parse_expr_snapshot!(expr_with_paren_2, r####"(2+3))"####);
parse_expr_snapshot!(expr_with_bracket_0, r####"[2,3]"####);
parse_expr_snapshot!(expr_with_bracket_1, r####"[[2,3]"####);
parse_expr_snapshot!(expr_with_bracket_2, r####"[2,3]]"####);
parse_expr_snapshot!(expr_with_bracket_3, r####"[2,3"####);
parse_expr_snapshot!(expr_with_bracket_4, r####"["####);
parse_expr_snapshot!(
    expr_with_bracket_5,
    r####"[
    1
    2,
]
        "####
);
parse_expr_snapshot!(
    expr_with_bracket_6,
    r####"[
    1,2,
]
        "####
);
parse_expr_snapshot!(
    expr_with_bracket_7,
    r####"[
    1,2,

        "####
);
parse_expr_snapshot!(expr_with_brace_0, r####"{a=2}"####);
parse_expr_snapshot!(expr_with_brace_1, r####"{a=2}}"####);
parse_expr_snapshot!(expr_with_delim_0, r####"({a=2}"####);
parse_expr_snapshot!(expr_with_delim_1, r####"({a=(2}"####);
parse_expr_snapshot!(expr_with_delim_2, r####"{a=[2]"####);
parse_expr_snapshot!(expr_with_delim_3, r####"[{a=2}"####);
parse_expr_snapshot!(expr_with_delim_4, r####"({a=[2}"####);
parse_expr_snapshot!(expr_with_delim_5, r####"{"####);
parse_expr_snapshot!(
    expr_with_delim_6,
    r####"{
    a = 1
}"####
);
parse_expr_snapshot!(
    expr_with_delim_7,
    r####"{
    a = 1
"####
);
