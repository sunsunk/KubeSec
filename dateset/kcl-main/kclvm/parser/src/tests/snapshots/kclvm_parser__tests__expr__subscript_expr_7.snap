---
source: parser/src/tests/expr.rs
expression: "crate::tests::parsing_expr_string(r####\"a[::]\"####)"
---
Node {
    node: Subscript(
        Subscript {
            value: Node {
                node: Identifier(
                    Identifier {
                        names: [
                            Node {
                                node: "a",
                                filename: "",
                                line: 1,
                                column: 0,
                                end_line: 1,
                                end_column: 1,
                            },
                        ],
                        pkgpath: "",
                        ctx: Load,
                    },
                ),
                filename: "",
                line: 1,
                column: 0,
                end_line: 1,
                end_column: 1,
            },
            index: None,
            lower: None,
            upper: None,
            step: None,
            ctx: Load,
            has_question: false,
        },
    ),
    filename: "",
    line: 1,
    column: 0,
    end_line: 1,
    end_column: 5,
}

